### PR TITLE
feat(graph): follow cross-service links in flow verdicts and signature deltas

### PR DESCRIPTION
**Pull request by vitali87**
Part of #1603, the second slice after #1975 (which added the cross-service edges' readers). Independent of it: both build on the `EXPOSES` / `RESOLVES_TO` edges already on `main`.

A flow walk ended at a client's `NETWORK` resource. The handler it resolved to lives in another project, whose `FLOWS_TO` edges were never loaded, so a sink reachable through the service boundary read as `NO_FLOW` with full coverage: a verified absence that was not one. `flow_reachability_verdict` now reads the remote edges of the whole graph (a `NETWORK` resource into the handler exposing the endpoint it resolves to; an RPC or dispatch resource into its handler directly), loads the flow edges of every handler project the hop leads into, and continues the walk. The pairs where the path crossed a service are returned as `FlowVerdict.remote_hops`, and the MCP `flow_verdict` handler maps them across as `remote_hops`. Coverage gaps stay the asked project's, since the verdict is asked of it. The handler's project is read off its qualified name, as `contract_linking` does: only `ENDPOINT` resources carry a `project` property, RPC and dispatch resources do not (local review P1, the test now crosses projects through an RPC resource).

The structural delta after a write lists, on every signature change, the `remote_callers`: call sites in any project that reach an endpoint the changed definition exposes, through a `NETWORK` resource or directly. No `CALLS` edge would ever name them, so `sites` alone said a handler's change was contained when it was not. Two queries (`RETURN DISTINCT`, so a client that reads and writes one URL is one caller; local review P2), emulated in `evals/cgr_graph.py` the way the real ones join.

Docs: `multi-project.md` (the two analyses that follow the link), the generated `mcp-server.md` tool table, `structural-delta.md`'s `signature_changes` row.

Tests: `test_flow_verdict.py` (a network hop into another project's handler, an RPC hop that crosses projects with no `project` property on the resource, a purely local path reporting no hops, the coverage-gaps read issued for the asked project only while the other project's edges are read under its own name), `test_structural_delta.py` (remote callers through a network resource and an RPC resource, with a read-and-write client counted once; the control: a change on a definition exposing nothing reports an empty list), `test_mcp_crash_correlation.py` (the handler returns `remote_hops`). Mutations, each restored: remote edges never added (network, RPC and MCP tests red), every pair reported as a hop (local-path test red), other projects' edges not loaded (network and gaps tests red), remote callers never read (delta test red), direct kinds ignored in the emulator (delta test red, RPC row missing), emulator not distinct (delta test red, duplicate row), project not derived from the handler (RPC, network and gaps tests red).

## Summary by CodeRabbit

- **New Features**
  - Flow analysis now follows paths across service boundaries, identifies remote hops, and accounts for coverage in every project entered.
  - Structural change reports include remote callers reaching changed service endpoints while excluding same-project callers.
  - MCP flow verdicts expose remote-hop details.

- **Bug Fixes**
  - Duplicate remote callers are now reported only once.

- **Documentation**
  - Updated guides and architecture documentation to describe cross-service flow tracing and remote caller reporting.

- **Tests**
  - Added coverage for remote flow paths, cross-service callers, deduplication, and MCP responses.

### CODE REVIEW

**Comment by vitali87** [comment]
@greptileai review

File: codebase_rag/flow_verdict.py [hunk 1]
```diff
@@ -73,15 +93,36 @@ def flow_reachability_verdict(
...
+        )
 
     if path := _bfs_path(edges, source_qn, sink_qn):
-        return FlowVerdict(FLOW_VERDICT_FOUND, tuple(path), ())
+        hops = tuple(
+            (a, b) for a, b in zip(path, path[1:], strict=False) if (a, b) in remote
+        )
+        return FlowVerdict(FLOW_VERDICT_FOUND, tuple(path), (), hops)
 
     gap_rows = fetch_all(CYPHER_FLOW_COVERAGE_GAPS, params)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/flow_verdict.py]: Fixed in 208d490f: coverage gaps are read for every project the walk entered (the asked one plus each handler project a reachable hop led into), so an uncovered module on the far side makes the verdict UNKNOWN. Test: no path, `q` entered and uncovered, `p` fully covered → UNKNOWN naming `q`'s module; gaps read for exactly `p` and `q`.

File: codebase_rag/flow_verdict.py [hunk 2]
```diff
@@ -73,15 +93,36 @@ def flow_reachability_verdict(
...
+                },
+            ),
+        )
 
     if path := _bfs_path(edges, source_qn, sink_qn):
-        return FlowVerdict(FLOW_VERDICT_FOUND, tuple(path), ())
+        hops = tuple(
+            (a, b) for a, b in zip(path, path[1:], strict=False) if (a, b) in remote
+        )
+        return FlowVerdict(FLOW_VERDICT_FOUND, tuple(path), (), hops)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/flow_verdict.py:143]: Fixed in 208d490f: a hop is recorded in `remote_hops` only when the handler's project differs from the project the walk was in before the resource. Test: a NETWORK resource resolving to a same-project handler is followed and reports no hop.

File: codebase_rag/cypher_queries.py [hunk 3]
```diff
@@ -786,6 +786,22 @@ def build_create_relationship_query(
...
+MATCH (c)-[r:READS_FROM|WRITES_TO]->(n:Resource)-[:RESOLVES_TO]->(e)
+RETURN DISTINCT h.qualified_name AS handler, e.name AS endpoint,
+       labels(c)[0] AS label, c.qualified_name AS qualified_name, c.path AS path,
+       n.name AS url"""
+CYPHER_DELTA_REMOTE_DIRECT_CALLERS_OF = """MATCH (h)-[:EXPOSES]->(e:Resource)
+WHERE h.qualified_name IN $qns
+MATCH (c)-[r:READS_FROM|WRITES_TO]->(e)
+RETURN DISTINCT h.qualified_name AS handler, e.name AS endpoint,
+       labels(c)[0] AS label, c.qualified_name AS qualified_name, c.path AS path,
+       e.name AS url"""
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/cypher_queries.py:807]: Fixed in 208d490f: both remote-caller queries exclude callers under the handler project's prefix (`NOT c.qualified_name STARTS WITH $project_prefix`), the emulator mirrors it, and the test plants a same-project caller over HTTP and over RPC that must not be listed.

File: codebase_rag/flow_verdict.py [hunk 4]
```diff
@@ -73,15 +93,36 @@ def flow_reachability_verdict(
...
+        _add_edges(
+            edges,
+            fetch_all(
+                CYPHER_FLOW_EDGES,
+                {
+                    cs.KEY_PROJECT_PREFIX: f"{other}{cs.SEPARATOR_DOT}",
+                    cs.KEY_PROJECT_NAME: other,
+                },
+            ),
+        )
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/flow_verdict.py]: Fixed in 208d490f: the remote read still returns edges only (one read), but a handler project's FLOWS_TO edges load only when a hop into it is reachable from the source, iterated to a fixpoint. Test: a local path with an unrelated hop into `q` in the graph reads edges for `p` alone; the entered-projects test shows `r`, reachable from nothing, is never read.

File: codebase_rag/cypher_queries.py [hunk 5]
```diff
@@ -786,6 +786,22 @@ def build_create_relationship_query(
...
+# dispatch kinds. Not project-scoped on the caller side, by design.
+CYPHER_DELTA_REMOTE_CALLERS_OF = """MATCH (h)-[:EXPOSES]->(e:Resource)
+WHERE h.qualified_name IN $qns
+MATCH (c)-[r:READS_FROM|WRITES_TO]->(n:Resource)-[:RESOLVES_TO]->(e)
+RETURN DISTINCT h.qualified_name AS handler, e.name AS endpoint,
+       labels(c)[0] AS label, c.qualified_name AS qualified_name, c.path AS path,
+       n.name AS url"""
+CYPHER_DELTA_REMOTE_DIRECT_CALLERS_OF = """MATCH (h)-[:EXPOSES]->(e:Resource)
+WHERE h.qualified_name IN $qns
+MATCH (c)-[r:READS_FROM|WRITES_TO]->(e)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/cypher_queries.py:803]: Fixed in 208d490f: same-project callers are excluded in both queries and the emulator; the delta test plants a same-project caller over HTTP and over RPC and asserts it is absent from `remote_callers`.

File: codebase_rag/flow_verdict.py [hunk 6]
```diff
@@ -73,15 +93,36 @@ def flow_reachability_verdict(
...
 
     if path := _bfs_path(edges, source_qn, sink_qn):
-        return FlowVerdict(FLOW_VERDICT_FOUND, tuple(path), ())
+        hops = tuple(
+            (a, b) for a, b in zip(path, path[1:], strict=False) if (a, b) in remote
+        )
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/flow_verdict.py:142]: Fixed in 208d490f: the edge stays in the walk for reachability and is a hop only when the project before the resource differs from the handler's. Tests: same-project NETWORK hop reports none; the RPC case crosses projects and reports one.

**Comment by vitali87** [comment]
@greptileai review

File: codebase_rag/structural_delta.py [hunk 7]
```diff
@@ -662,14 +676,59 @@ def _arity_findings(after: Snapshot, repo_root: Path | None) -> list[ArityAtSite
...
+    if not handlers:
+        return found
+    for query in (
+        cq.CYPHER_DELTA_REMOTE_CALLERS_OF,
+        cq.CYPHER_DELTA_REMOTE_DIRECT_CALLERS_OF,
+    ):
+        for row in fetch_all(
+            query,
+            {cs.KEY_QNS: handlers, cs.KEY_PROJECT_PREFIX: _prefix(project_name)},
+        ):
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/structural_delta.py:697]: Fixed in fd1a8f58: rows from the two shapes are deduplicated on the full row (handler and every caller field), so a caller both return is listed once while two callers sharing a name but not a URL both survive; unit test with a fake read returning the same row for both queries.

**Comment by vitali87** [comment]
@greptileai review

**Comment by vitali87** [comment]
@greptileai review

**Comment by vitali87** [comment]
Adopted as orphaned (the session that created it is offline, and the last commit is about 25h old) and rebased onto main with --force-with-lease. Rebased rather than merged because main's ruleset has require_extra_approval_for_unattributed_changes.

- Conflict: `constants/graph.py` KEY_* block. Main already carried a superset (KEY_DIRECTION, KEY_CALLERS, KEY_HANDLER_PROJECT, from 5215cca1), so main's side is kept; there are no duplicate names.
- Checked that `_delta_remote_callers_of` in `evals/cgr_graph.py` works with main's site-keyed edges: it reads only fields 0-4 and removes duplicate rows by content.
- Local: test_crash_correlation, test_flow_verdict, test_mcp_crash_correlation and test_structural_delta pass, 98 passed / 0 skipped.

File: codebase_rag/cypher_queries.py [hunk 8]
```diff
@@ -786,6 +786,25 @@ def build_create_relationship_query(
...
+WHERE NOT c.qualified_name STARTS WITH $project_prefix
+RETURN DISTINCT h.qualified_name AS handler, e.name AS endpoint,
+       labels(c)[0] AS label, c.qualified_name AS qualified_name, c.path AS path,
+       e.name AS url"""
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/cypher_queries.py]: Already in place: both queries enforce the kinds. `CYPHER_DELTA_REMOTE_CALLERS_OF` matches `(n:Resource {kind: 'NETWORK'})` on the RESOLVES_TO hop, and `CYPHER_DELTA_REMOTE_DIRECT_CALLERS_OF` filters `e.kind IN ['RPC', 'DISPATCH']`. A later commit on this branch added them, the same filters `CYPHER_FLOW_REMOTE_EDGES` uses.

File: codebase_rag/flow_verdict.py [hunk 9]
```diff
@@ -73,27 +95,95 @@ def flow_reachability_verdict(
...
+    # Resource-to-handler hops join the graph up front (one read of edges
+    # only); a handler project's own flow edges load when the walk reaches
+    # a hop into it, and not before, so a local verdict never reads an
+    # unrelated project (issue #1603; bot review on PR #1978).
+    remote: dict[str, set[str]] = {}
+    for row in fetch_all(CYPHER_FLOW_REMOTE_EDGES, None):
         source, target = row.get("source"), row.get("target")
         if isinstance(source, str) and isinstance(target, str):
             edges.setdefault(source, []).append(target)
+            remote.setdefault(source, set()).add(target)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/flow_verdict.py]: Leaving as is. `CYPHER_FLOW_REMOTE_EDGES` reads only the cross-service hops (NETWORK→endpoint→handler, and RPC/DISPATCH→handler), one row per exposed endpoint, not the flow graph. That set grows with the number of endpoints, not with code size, and the per-project FLOWS_TO edges, which are the large part, already load lazily as the walk enters each project. If profiling on a large multi-service graph shows it matters, scoping by reachable resources is a clean follow-up, but it isn't a correctness issue.

File: codebase_rag/structural_delta.py [hunk 10]
```diff
@@ -662,14 +676,71 @@ def _arity_findings(after: Snapshot, repo_root: Path | None) -> list[ArityAtSite
...
+            handler = _text(row.get(cs.KEY_HANDLER))
+            caller = RemoteCaller(
+                qualified_name=_text(row.get(cs.KEY_QUALIFIED_NAME)),
+                label=_text(row.get(cs.KEY_LABEL)),
+                path=_text(row.get(cs.KEY_PATH)),
+                url=_text(row.get(cs.KEY_URL)),
+                endpoint=_text(row.get(cs.KEY_ENDPOINT)),
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/structural_delta.py]: Not reachable: `_text` in `structural_delta.py` is `str(value) if isinstance(value, str) else ""`, which never returns `None`. A missing `path`/`url`/`endpoint` becomes `""`, so `RemoteCaller`'s `str` fields hold.

File: codebase_rag/flow_verdict.py [hunk 11]
```diff
@@ -73,27 +95,95 @@ def flow_reachability_verdict(
...
+                edges,
+                fetch_all(
+                    CYPHER_FLOW_EDGES,
+                    {
+                        cs.KEY_PROJECT_PREFIX: f"{other}{cs.SEPARATOR_DOT}",
+                        cs.KEY_PROJECT_NAME: other,
+                    },
+                ),
+            )
+        loaded |= entered
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/flow_verdict.py]: Leaving as is. The loop runs once per *round* of newly entered projects, bounded by the number of projects the walk crosses into (usually one or two), and each round's reachability is linear in the edges loaded so far. Folding project loading into the BFS would save the re-walk, at the cost of interleaving I/O with traversal. That's worth it only if a real graph shows many rounds; I'm keeping the simpler, clearly correct form.

File: codebase_rag/flow_verdict.py [hunk 12]
```diff
@@ -73,27 +95,95 @@ def flow_reachability_verdict(
...
         )
-    )
     if gaps:
-        return FlowVerdict(FLOW_VERDICT_UNKNOWN, (), gaps)
+        return FlowVerdict(FLOW_VERDICT_UNKNOWN, (), tuple(sorted(gaps)))
     return FlowVerdict(FLOW_VERDICT_NO_FLOW, (), ())
 
 
+def _project_of(qn: str) -> str:
+    return qn.split(cs.SEPARATOR_DOT, 1)[0]
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/flow_verdict.py]: Fixed in 96133ad9, and the real defect was wider than the missing-separator case: `_project_of` took the FIRST dotted segment, so a handler in project `svc.v2` was attributed to `svc`. The walk loaded the wrong project's edges, and the remote-hop test compared the wrong projects. It now uses the longest registered project name the qn sits under, reading the registry once (the #1970 rule), and falls back to the first segment only when no registered project owns it. `test_a_handler_in_a_dotted_project_loads_that_projects_edges` fails on the old code.

File: codebase_rag/structural_delta.py [hunk 13]
```diff
@@ -662,14 +676,71 @@ def _arity_findings(after: Snapshot, repo_root: Path | None) -> list[ArityAtSite
...
+                caller["path"],
+                caller["url"],
+                caller["endpoint"],
+            )
+            if handler and key not in seen:
+                seen.add(key)
+                found.setdefault(handler, []).append(caller)
+    for rows in found.values():
+        rows.sort(key=lambda r: (r["qualified_name"], r["url"], r["path"]))
+    return found
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/structural_delta.py]: Not reachable, same as above: `_text` returns `""` for a missing or non-string value, never `None`, so every `RemoteCaller` field and the sort key are strings.
